File: js_ast_parser.js
#!/usr/bin/env node
/**
 * JavaScript AST Parser using Babel
 * Parses ES6+ JavaScript files and generates AST in XML format
 */

const fs = require('fs');
const path = require('path');
const parser = require('@babel/parser');
const traverse = require('@babel/traverse').default;

/**
 * Parse JavaScript file and generate AST
 * @param {string} filePath - Path to JavaScript file
 * @param {string} outputPath - Path to output XML file
 */
function parseJavaScriptToAST(filePath, outputPath) {
  try {
    // Read source code
    const sourceCode = fs.readFileSync(filePath, 'utf-8');
    
    // Parse with Babel
    const ast = parser.parse(sourceCode, {
      sourceType: 'module',
      plugins: [
        'jsx',
        'classProperties',
        'decorators-legacy',
        'asyncGenerators',
        'bigInt',
        'dynamicImport',
        'nullishCoalescingOperator',
        'optionalChaining'
      ]
    });
    
    // Extract metadata
    const metadata = {
      imports: [],
      exports: [],
      classes: [],
      functions: [],
      variables: []
    };
    
    // Traverse AST to extract information
    traverse(ast, {
      ImportDeclaration(path) {
        const importInfo = {
          source: path.node.source.value,
          specifiers: path.node.specifiers.map(spec => {
            if (spec.type === 'ImportDefaultSpecifier') {
              return { type: 'default', local: spec.local.name };
            } else if (spec.type === 'ImportSpecifier') {
              return { 
                type: 'named', 
                imported: spec.imported.name, 
                local: spec.local.name 
              };
            }
            return null;
          }).filter(Boolean)
        };
        metadata.imports.push(importInfo);
      },
      
      ExportDefaultDeclaration(path) {
        metadata.exports.push({
          type: 'default',
          name: path.node.declaration.id?.name || 'anonymous'
        });
      },
      
      ExportNamedDeclaration(path) {
        if (path.node.declaration) {
          if (path.node.declaration.type === 'ClassDeclaration') {
            metadata.exports.push({
              type: 'class',
              name: path.node.declaration.id.name
            });
          } else if (path.node.declaration.type === 'FunctionDeclaration') {
            metadata.exports.push({
              type: 'function',
              name: path.node.declaration.id.name
            });
          }
        }
      },
      
      ClassDeclaration(path) {
        const classInfo = {
          name: path.node.id.name,
          superClass: path.node.superClass?.name || null,
          methods: []
        };
        
        path.node.body.body.forEach(member => {
          if (member.type === 'ClassMethod') {
            classInfo.methods.push({
              name: member.key.name,
              kind: member.kind, // 'constructor', 'method', 'get', 'set'
              async: member.async,
              static: member.static,
              params: member.params.map(p => p.name || p.type)
            });
          } else if (member.type === 'ClassProperty') {
            // Class properties
            if (!classInfo.properties) {
              classInfo.properties = [];
            }
            classInfo.properties.push({
              name: member.key.name,
              static: member.static
            });
          }
        });
        
        metadata.classes.push(classInfo);
      },
      
      FunctionDeclaration(path) {
        // Skip if it's inside a class (already captured)
        if (path.parent.type === 'Program' || path.parent.type === 'ExportNamedDeclaration') {
          metadata.functions.push({
            name: path.node.id.name,
            async: path.node.async,
            params: path.node.params.map(p => p.name || p.type)
          });
        }
      }
    });
    
    // Generate XML output
    const xml = generateXML(metadata, path.basename(filePath));
    
    // Write to output file
    fs.writeFileSync(outputPath, xml, 'utf-8');
    
    console.log(`AST generated successfully: ${outputPath}`);
    return true;
  } catch (error) {
    console.error(`Failed to parse ${filePath}:`, error.message);
    return false;
  }
}

/**
 * Generate XML from metadata
 * @param {object} metadata - Extracted metadata
 * @param {string} fileName - Original file name
 * @returns {string} XML string
 */
function generateXML(metadata, fileName) {
  let xml = '<?xml version="1.0" encoding="UTF-8"?>\n';
  xml += `<JavaScriptFile name="${fileName}">\n`;
  
  // Imports
  if (metadata.imports.length > 0) {
    xml += '  <Imports>\n';
    metadata.imports.forEach(imp => {
      xml += `    <Import source="${escapeXml(imp.source)}">\n`;
      imp.specifiers.forEach(spec => {
        xml += `      <Specifier type="${spec.type}" imported="${escapeXml(spec.imported || spec.local)}" local="${escapeXml(spec.local)}" />\n`;
      });
      xml += '    </Import>\n';
    });
    xml += '  </Imports>\n';
  }
  
  // Exports
  if (metadata.exports.length > 0) {
    xml += '  <Exports>\n';
    metadata.exports.forEach(exp => {
      xml += `    <Export type="${exp.type}" name="${escapeXml(exp.name)}" />\n`;
    });
    xml += '  </Exports>\n';
  }
  
  // Classes
  if (metadata.classes.length > 0) {
    xml += '  <Classes>\n';
    metadata.classes.forEach(cls => {
      xml += `    <Class name="${escapeXml(cls.name)}"`;
      if (cls.superClass) {
        xml += ` superClass="${escapeXml(cls.superClass)}"`;
      }
      xml += '>\n';
      
      // Properties
      if (cls.properties && cls.properties.length > 0) {
        xml += '      <Properties>\n';
        cls.properties.forEach(prop => {
          xml += `        <Property name="${escapeXml(prop.name)}" static="${prop.static}" />\n`;
        });
        xml += '      </Properties>\n';
      }
      
      // Methods
      if (cls.methods.length > 0) {
        xml += '      <Methods>\n';
        cls.methods.forEach(method => {
          xml += `        <Method name="${escapeXml(method.name)}" kind="${method.kind}" async="${method.async}" static="${method.static}">\n`;
          if (method.params.length > 0) {
            xml += '          <Parameters>\n';
            method.params.forEach(param => {
              xml += `            <Parameter name="${escapeXml(param)}" />\n`;
            });
            xml += '          </Parameters>\n';
          }
          xml += '        </Method>\n';
        });
        xml += '      </Methods>\n';
      }
      
      xml += '    </Class>\n';
    });
    xml += '  </Classes>\n';
  }
  
  // Functions
  if (metadata.functions.length > 0) {
    xml += '  <Functions>\n';
    metadata.functions.forEach(func => {
      xml += `    <Function name="${escapeXml(func.name)}" async="${func.async}">\n`;
      if (func.params.length > 0) {
        xml += '      <Parameters>\n';
        func.params.forEach(param => {
          xml += `        <Parameter name="${escapeXml(param)}" />\n`;
        });
        xml += '      </Parameters>\n';
      }
      xml += '    </Function>\n';
    });
    xml += '  </Functions>\n';
  }
  
  xml += '</JavaScriptFile>\n';
  return xml;
}

/**
 * Escape XML special characters
 * @param {string} str - String to escape
 * @returns {string} Escaped string
 */
function escapeXml(str) {
  if (!str) return '';
  return String(str)
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&apos;');
}

// Command line interface
if (require.main === module) {
  const args = process.argv.slice(2);
  
  if (args.length < 2) {
    console.error('Usage: node js_ast_parser.js <input.js> <output.xml>');
    process.exit(1);
  }
  
  const inputFile = args[0];
  const outputFile = args[1];
  
  if (!fs.existsSync(inputFile)) {
    console.error(`Input file not found: ${inputFile}`);
    process.exit(1);
  }
  
  const success = parseJavaScriptToAST(inputFile, outputFile);
  process.exit(success ? 0 : 1);
}

module.exports = { parseJavaScriptToAST };

File: backend/ast_api/js_ast_parser.py
"""
JavaScript AST Parser
Parse Babel-generated JavaScript AST XML files
"""
import xml.etree.ElementTree as ET
from pathlib import Path
import logging

logger = logging.getLogger(__name__)


class JavaScriptASTParser:
    """JavaScript AST XML Parser"""
    
    def __init__(self, ast_file_path):
        self.file_path = Path(ast_file_path)
        self.tree = None
        self.root = None
        self.component_data = {}
        
    def parse(self):
        """Parse JavaScript AST file"""
        try:
            self.tree = ET.parse(self.file_path)
            self.root = self.tree.getroot()
            
            # Extract component info
            self.component_data = self._extract_component_info()
            
            # Extract imports
            self.component_data['imports'] = self._extract_imports()
            
            # Extract exports
            self.component_data['exports'] = self._extract_exports()
            
            # Extract classes
            self.component_data['classes'] = self._extract_classes()
            
            # Extract functions
            self.component_data['functions'] = self._extract_functions()
            
            return self.component_data
            
        except Exception as e:
            logger.error(f"Failed to parse JavaScript AST file {self.file_path}: {e}")
            raise
    
    def _extract_component_info(self):
        """Extract basic component information"""
        return {
            'name': self.root.get('name', 'Unknown').replace('.js', ''),
            'type': 'LWCComponent',
            'fileName': self.file_path.name,
        }
    
    def _extract_imports(self):
        """Extract import statements"""
        imports = []
        imports_elem = self.root.find('Imports')
        
        if imports_elem is not None:
            for imp in imports_elem.findall('Import'):
                import_info = {
                    'source': imp.get('source', ''),
                    'specifiers': []
                }
                
                # 提取specifier的名称（imported或local）
                for spec in imp.findall('Specifier'):
                    imported = spec.get('imported', '')
                    local = spec.get('local', '')
                    # 使用local名称（如果不同）或imported名称
                    name = local if local else imported
                    if name:
                        import_info['specifiers'].append(name)
                
                # 检测Apex依赖关系
                source = import_info['source']
                if source.startswith('@salesforce/apex/'):
                    import_info['apex_dependency'] = self._parse_apex_import(source, import_info['specifiers'])
                
                imports.append(import_info)
        
        return imports
    
    def _parse_apex_import(self, source, specifiers):
        """
        解析Apex导入，提取类名和方法名
        例如: '@salesforce/apex/PropertyController.getPagedPropertyList'
        """
        apex_info = {
            'type': 'apex',
            'class_name': None,
            'method_name': None,
            'full_path': source
        }
        
        # 移除 @salesforce/apex/ 前缀
        apex_path = source.replace('@salesforce/apex/', '')
        
        # 解析类名和方法名
        if '.' in apex_path:
            parts = apex_path.split('.')
            apex_info['class_name'] = parts[0]
            if len(parts) > 1:
                apex_info['method_name'] = parts[1]
        else:
            apex_info['class_name'] = apex_path
        
        return apex_info
    
    def _extract_exports(self):
        """Extract export statements"""
        exports = []
        exports_elem = self.root.find('Exports')
        
        if exports_elem is not None:
            for exp in exports_elem.findall('Export'):
                exports.append({
                    'type': exp.get('type', ''),
                    'name': exp.get('name', '')
                })
        
        return exports
    
    def _extract_classes(self):
        """Extract class definitions"""
        classes = []
        classes_elem = self.root.find('Classes')
        
        if classes_elem is not None:
            for cls in classes_elem.findall('Class'):
                class_info = {
                    'name': cls.get('name', ''),
                    'superClass': cls.get('superClass', ''),
                    'properties': [],
                    'methods': []
                }
                
                # Extract properties
                properties_elem = cls.find('Properties')
                if properties_elem is not None:
                    for prop in properties_elem.findall('Property'):
                        class_info['properties'].append({
                            'name': prop.get('name', ''),
                            'static': prop.get('static', 'false') == 'true'
                        })
                
                # Extract methods
                methods_elem = cls.find('Methods')
                if methods_elem is not None:
                    for method in methods_elem.findall('Method'):
                        method_info = {
                            'name': method.get('name', ''),
                            'kind': method.get('kind', 'method'),
                            'async': method.get('async', 'false') == 'true',
                            'static': method.get('static', 'false') == 'true',
                            'parameters': []
                        }
                        
                        # Extract parameters
                        params_elem = method.find('Parameters')
                        if params_elem is not None:
                            for param in params_elem.findall('Parameter'):
                                method_info['parameters'].append(param.get('name', ''))
                        
                        class_info['methods'].append(method_info)
                
                classes.append(class_info)
        
        return classes
    
    def _extract_functions(self):
        """Extract function definitions"""
        functions = []
        functions_elem = self.root.find('Functions')
        
        if functions_elem is not None:
            for func in functions_elem.findall('Function'):
                func_info = {
                    'name': func.get('name', ''),
                    'async': func.get('async', 'false') == 'true',
                    'parameters': []
                }
                
                # Extract parameters
                params_elem = func.find('Parameters')
                if params_elem is not None:
                    for param in params_elem.findall('Parameter'):
                        func_info['parameters'].append(param.get('name', ''))
                
                functions.append(func_info)
        
        return functions


def parse_js_ast_file(file_path):
    """
    Parse JavaScript AST XML file
    
    Args:
        file_path: Path to JavaScript AST XML file
        
    Returns:
        Dictionary containing parsed component data
    """
    parser = JavaScriptASTParser(file_path)
    return parser.parse()
